api/agent/async.go

```diff
@@ -35,15 +34,20 @@ func (a *agent) asyncDequeue() {
 
 		// we think we can get a cookie now, so go get a cookie
 		select {
-		case <-a.shutdown:
+		case <-a.shutWg.Closer():
+			a.shutWg.AddSession(-1)
 			return
 		case model, ok := <-a.asyncChew(ctx):
 			if ok {
-				a.wg.Add(1) // need to add 1 in this thread to ensure safe shutdown
 				go func(model *models.Call) {
 					a.asyncRun(ctx, model)
-					a.wg.Done() // can shed it after this is done, Submit will add 1 too but it's fine
+					a.shutWg.AddSession(-1)
 				}(model)
+
+				// WARNING: tricky. We reserve another session for next iteration of the loop
```

Comment:
why? there was a waiter covering the entire asyncDequeue, did we remove it?

Reply:
so for-loop maintains the original +1 (from agent.go). If asyncRun go routine is spawned, then we need one more. With this, instead of defer -1, we've got for-loop return -1 in select statements. A bit hard to see.

Reply:
It basically avoids a failure case. After asyncChew, we would need to +1, but if it fails, then we drop the dequeued item, but we avoid that here by using the for-loop +1, then +1 another one after asyncRun() launch.